src/FakeItEasy/Creation/LoopDetectingResolutionContext.cs

```diff
@@ -20,10 +20,10 @@ internal class LoopDetectingResolutionContext
         public bool TryBeginToResolve(Type type) => this.typesCurrentlyBeingResolved.Add(type);
 
         /// <summary>
-        /// Indicate successful resolution of a <paramref name="type"/> object.
+        /// Indicate the end of an attempt to resolve a <paramref name="type"/> object.
         /// Subsequent calls to <see cref="TryBeginToResolve"/> will now pass.
         /// </summary>
         /// <param name="type">The type that was resolved.</param>
-        public void OnSuccessfulResolve(Type type) => this.typesCurrentlyBeingResolved.Remove(type);
+        public void EndResolve(Type type) => this.typesCurrentlyBeingResolved.Remove(type);

```

Comment:
The changes to this file were not strictly required to fix the issue, but for a brief time, I forgot to call this after I successfully used a cached resolver strategy. Then I thought more about this and realized that, like refraining from trapping argument constraints, we should do this every time we resolve a Dummy.
Failing to remove the "in progress" marker from this container might mean that we later throw an exception indicating a recursive resolution when it's not actually happening. I think we'd need non-deterministic constructors in the classes for this to become an issue, but it's easy to guard against.

Reply:
Good thinking.